source/text_handler.cpp

```diff
@@ -229,25 +229,24 @@ spv_result_t AssemblyContext::binaryEncodeU64(const uint64_t value,
 spv_result_t AssemblyContext::binaryEncodeNumericLiteral(
     const char* val, spv_result_t error_code, const IdType& type,
     spv_instruction_t* pInst) {
-  const bool is_bottom = type.type_class == libspirv::IdTypeClass::kBottom;
-  const bool is_floating = libspirv::isScalarFloating(type);
-  const bool is_integer = libspirv::isScalarIntegral(type);
-
-  if (!is_bottom && !is_floating && !is_integer) {
-    return diagnostic(SPV_ERROR_INTERNAL)
-           << "The expected type is not a scalar integer or float type";
+  using spvutils::EncodeNumberStatus;
+  EncodeNumberStatus rc = EncodeNumberStatus::kSuccess;
+  std::string error_msg;
+  std::tie(rc, error_msg) = ParseAndEncodeNumber(
+      val, type,
+      [this, pInst](uint32_t d) { this->binaryEncodeU32(d, pInst); });
+  switch (rc) {
+    case EncodeNumberStatus::kSuccess:
+      return SPV_SUCCESS;
+    case EncodeNumberStatus::kInvalidText:
+      return diagnostic(error_code) << error_msg;
+    case EncodeNumberStatus::kUnsupported:
+      return diagnostic(SPV_ERROR_INTERNAL) << error_msg;
+    case EncodeNumberStatus::kInvalidUsage:
+      return diagnostic(SPV_ERROR_INVALID_TEXT) << error_msg;
   }
-
-  // If this is bottom, but looks like a float, we should treat it like a
-  // float.
-  const bool looks_like_float = is_bottom && strchr(val, '.');
-
-  // If we explicitly expect a floating-point number, we should handle that
-  // first.
-  if (is_floating || looks_like_float)
-    return binaryEncodeFloatingPointLiteral(val, error_code, type, pInst);
-
-  return binaryEncodeIntegerLiteral(val, error_code, type, pInst);
+  // This line is not reachable, only added to stasify the compiler.
+  return SPV_SUCCESS;
```

Comment:
Better to return ERROR_INTERNAL with a descriptive message like "unexpected result code from ..."